File: sol/solution/0201-0300/0203/Solution.ex
# Definition for singly-linked list.
#
# defmodule ListNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           next: ListNode.t() | nil
#         }
#   defstruct val: 0, next: nil
# end
defmodule Solution do
  @spec remove_elements(head :: ListNode.t | nil, val :: integer) :: ListNode.t | nil
  def remove_elements(nil, _val), do: nil
  def remove_elements(%ListNode{val: val, next: next}, val) do
    remove_elements(next, val)
  end
  def remove_elements(%ListNode{val: val, next: next}, target) do
    %ListNode{val: val, next: remove_elements(next, target)}
  end
end

File: sol/solution/0201-0300/0203/Solution.erl
%% Definition for singly-linked list.
%%
%% -record(list_node, {val = 0 :: integer(),
%%                     next = null :: 'null' | #list_node{}}).
-spec remove_elements(Head :: #list_node{} | null, Val :: integer()) -> #list_node{} | null.
remove_elements(null, _) ->
    null;
remove_elements(#list_node{val = Val, next = Next}, Val) ->
    remove_elements(Next, Val);
remove_elements(#list_node{val = H, next = T}, Val) ->
    #list_node{val = H, next = remove_elements(T, Val)}.